### PR TITLE
LibJS/Temporal Reorder and Unroll Accesses to Fields

### PR DESCRIPTION
**Pull request by Epigenetic**
This ticks three boxes in #15525:
https://github.com/tc39/proposal-temporal/commit/f2d564238f79b2664b00026741d698b9e46fda31
https://github.com/tc39/proposal-temporal/commit/07673d3dfc8aab49164893ce958a033986e73978
https://github.com/tc39/proposal-temporal/commit/8e80575ad918a0f31b9e888e856f198227722ab0
Note that this last one seems to supersede some of the changes from the other revisions.

### CODE REVIEW

**Review by linusg — CHANGES_REQUESTED** [review]
Please always link to the upstream commits. See any previous temporal changes for examples.

File: Userland/Libraries/LibJS/Runtime/Value.cpp [hunk 1]
```diff
@@ -919,6 +919,24 @@ double to_integer_or_infinity(double number)
     return integer;
 }
 
+// https://tc39.es/proposal-temporal/#sec-tointegerifintegral
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Value.cpp:922]: This doesn't belong here as long as it's part of the Temporal proposal, see `Temporal/AbstractOperations.cpp/h`

**Comment by Epigenetic** [comment]
> Please always link to the upstream commits. See any previous temporal changes for examples.

Wasn't too sure which commit to reference for https://github.com/SerenityOS/serenity/pull/16057/commits/bd80e83c473765e223070365a9a299646e08e8da, referenced the one that changed the function name, not sure if I should also mention https://github.com/tc39/proposal-temporal/commit/7fbdd28238ee265c938f0675be79e3d68d2ebaf1 which added that Abstract Operation under a different name (and is not fully addressed here).

File: Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp [hunk 2]
```diff
@@ -165,7 +165,7 @@ ThrowCompletionOr<DurationRecord> to_temporal_duration_record(VM& vm, Value temp
     auto duration_record_fields = temporal_duration_record_fields<DurationRecord, double>(vm);
     auto partial_duration_record_fields = temporal_duration_record_fields<PartialDurationRecord, Optional<double>>(vm);
 
-    // 5. For each row of Table 8, except the header row, in table order, do
+    // 5. For each row of Table 8, except the header row, do
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp]: Missing changes from https://github.com/tc39/proposal-temporal/commit/f2d5642#diff-ceec2609082b71fe76df3f43d148b97d667dde310f51a514141886246bf3fc74R1048-R1057

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp [hunk 3]
```diff
@@ -1870,4 +1870,22 @@ ThrowCompletionOr<DifferenceSettings> get_difference_settings(VM& vm, Difference
     };
 }
 
+// https://tc39.es/proposal-temporal/#sec-tointegerifintegral
+ThrowCompletionOr<double> to_integer_if_integral(VM& vm, Value argument, StringView unit)
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp:1874]: The commit you linked is a plain rename, so do that instead?

File: Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp [hunk 4]
```diff
@@ -289,37 +289,81 @@ ThrowCompletionOr<PartialDurationRecord> to_temporal_partial_duration_record(VM&
     // 2. Let result be a new partial Duration Record with each field set to undefined.
     auto result = PartialDurationRecord {};
 
-    // 3. Let any be false.
-    auto any = false;
+    // 3. Let days be ? Get(temporalDurationLike, "days").
+    auto days = TRY(temporal_duration_like.as_object().get(PropertyKey(vm.names.days)));
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp:317]: `get(vm.names.days)`, same below

File: Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp [hunk 5]
```diff
@@ -318,70 +318,70 @@ ThrowCompletionOr<PartialDurationRecord> to_temporal_partial_duration_record(VM&
 
     // 4. If days is not undefined, set result.[[Days]] to ? ToIntegerIfIntegral(days).
     if (!days.is_undefined())
-        result.days = TRY(days.to_integer_if_integral(vm, "days"sv));
+        result.days = TRY(to_integer_if_integral(vm, days, "days"sv));
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp:321]: Don't make fixes like this across commits

File: Userland/Libraries/LibJS/Runtime/Temporal/PlainTime.cpp [hunk 6]
```diff
@@ -345,32 +345,89 @@ ThrowCompletionOr<TemporalTimeLikeRecord> to_temporal_time_record(VM& vm, Object
     auto* partial = TRY(prepare_temporal_fields(vm, temporal_time_like, { "hour"sv, "microsecond"sv, "millisecond"sv, "minute"sv, "nanosecond"sv, "second"sv }, PrepareTemporalFieldsPartial {}));
 
     // 3. Let result be a new TemporalTimeLike Record with each field set to undefined.
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/PlainTime.cpp:347]: Leftover comment

File: Userland/Libraries/LibJS/Runtime/Temporal/PlainTime.cpp [hunk 7]
```diff
@@ -345,32 +345,89 @@ ThrowCompletionOr<TemporalTimeLikeRecord> to_temporal_time_record(VM& vm, Object
...
-    auto result = TemporalTimeLikeRecord {};
+    TemporalTimeLikeRecord result;
+    // 3. If completeness is complete, then
+    if (completeness == ToTemporalTimeRecordCompleteness::Complete)
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/PlainTime.cpp:350]: If/else with more than one line needs curly braces

File: Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp [hunk 8]
```diff
@@ -165,29 +165,53 @@ ThrowCompletionOr<DurationRecord> to_temporal_duration_record(VM& vm, Value temp
     auto duration_record_fields = temporal_duration_record_fields<DurationRecord, double>(vm);
     auto partial_duration_record_fields = temporal_duration_record_fields<PartialDurationRecord, Optional<double>>(vm);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp:166]: Now unused

**Review by linusg — CHANGES_REQUESTED** [review]
You forgot to remove `temporal_duration_record_fields` / `temporal_time_like_record_fields`

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp (hunk 3 above)
Comment (Epigenetic) [reply Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp:1874]: The function with the prior rename (ToIntegerWithoutRounding ) did not exist previously, so there was no function to rename.  If its preferable, I could make a commit adding the function under the old name and then rename it in a second commit.

Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp:1874]: No idea what you're talkking about
https://github.com/SerenityOS/serenity/blob/f5a3355639a826d041b8b841e5525397677b42ee/Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.h#L204-L221

Comment (Epigenetic) [reply Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.cpp:1874]: Huh, don't know how I missed that

File: Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp [hunk 9]
```diff
@@ -292,7 +292,14 @@ ThrowCompletionOr<PartialDurationRecord> to_temporal_partial_duration_record(VM&
     // 3. Let any be false.
     auto any = false;
 
-    // 4. For each row of Table 7, except the header row, in table order, do
+    // 4. Let propertyNames be a new empty List.
+
+    // 5. For each row of Table 7, except the header row, do
+    // a. Append the property name in the Property Name column of the row to propertyNames.
+    // 6. Let sortedPropertyNames be SortStringListByCodeUnit(propertyNames).
+    // 7. For each property name property of sortedPropertyNames, do
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/Duration.cpp]: You can't just add comments here and then not implement the change...

File: Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.h [hunk 10]
```diff
@@ -201,9 +201,9 @@ ThrowCompletionOr<double> to_integer_throw_on_infinity(VM& vm, Value argument, E
     return integer;
 }
 
-// 13.41 ToIntegerWithoutRounding ( argument ), https://tc39.es/proposal-temporal/#sec-temporal-tointegerwithoutrounding
+// 13.41 ToIntegerWithoutRounding ( argument ), https://tc39.es/proposal-temporal/#sec-tointegerifintegral
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/AbstractOperations.h:204]: Function name not updated

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibJS/Runtime/Temporal/Duration.h [hunk 11]
```diff
@@ -121,16 +121,15 @@ template<typename StructT, typename ValueT>
 auto temporal_duration_record_fields = [](VM& vm) {
     using FieldT = TemporalDurationRecordField<StructT, ValueT>;
     return AK::Array {
+        FieldT { &StructT::years, vm.names.years },
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/Duration.h:124]: Unused, as per my previous comment

File: Userland/Libraries/LibJS/Runtime/Temporal/PlainTime.h [hunk 12]
```diff
@@ -77,11 +77,11 @@ auto temporal_time_like_record_fields = [](VM& vm) {
...
     return AK::Array {
         FieldT { &StructT::hour, vm.names.hour },
-        FieldT { &StructT::microsecond, vm.names.microsecond },
-        FieldT { &StructT::millisecond, vm.names.millisecond },
         FieldT { &StructT::minute, vm.names.minute },
-        FieldT { &StructT::nanosecond, vm.names.nanosecond },
         FieldT { &StructT::second, vm.names.second },
+        FieldT { &StructT::millisecond, vm.names.millisecond },
+        FieldT { &StructT::microsecond, vm.names.microsecond },
+        FieldT { &StructT::nanosecond, vm.names.nanosecond },
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/PlainTime.h:84]: Likewise

**Review by linusg — APPROVED** [review]
Thanks!

**Pull request merged**